[Clang][Sema] Do not perform type conversion before computing result type of `x ? : y` in C++

zwuis opened the pull request:
Currently, Clang performs some type conversions on the left operator of `x ? : y` before computing the result type of the operator. But, these conversions are extranous if the operands satisfy the conditions in [[expr.cond]/5](https://eel.is/c++draft/expr.cond#5) or [[expr.cond]/7.1](https://eel.is/c++draft/expr.cond#7.1).

```cpp
enum Enum { Eval } e;
struct S { operator short(); };

// Current:
// type of operands: Enum Enum
// after type conversions: int Enum
// result type: int
// Expected result type: Enum
using Case1 = decltype(e ? : Eval);
// Current:
// type of operands: short S
// after type conversions: int S
// result type: int
// Expected result type: short
using Case2 = decltype((short)0 ? : S{});
```

This PR removes these type conversions. Fixes #15998.

Drive-by changes: Move some tests to 'clang/test/CXX/drs/cwg5xx.cpp'

zwuis committed:
Fix computing result type of conditional operand

zwuis committed:
Format code

shafik commented on clang/test/SemaCXX/conditional-gnu-ext.cpp:
```diff
@@ -0,0 +1,19 @@
+// RUN: %clang_cc1 -fsyntax-only -verify -std=c++11 %s
```
I think this test should go in `clang/test/Sema/conditional-expr.c` and we should add a section of the GNU extension there.

I am also a bit concerned that we don't have a specific set of sema tests for the GNU form of the conditional expression. So it makes me much left confident that we will catch any breaks with this change.

zwuis replied:
> I think this test should go in `clang/test/Sema/conditional-expr.c` and we should add a section of the GNU extension there.
> 
> I am also a bit concerned that we don't have a specific set of sema tests for the GNU form of the conditional expression. So it makes me much left confident that we will catch any breaks with this change.

I copied most test cases from 'clang/test/Sema/conditional-expr.c' and 'clang/test/SemaCXX/conditional-expr.cpp' to new files, and modified them to FNU extension form.

Please let me know if the test cases still should be merged into one file or we need other test cases.

erichkeane commented on clang/lib/Sema/SemaExpr.cpp:
```diff
@@ -8785,14 +8785,11 @@ ExprResult Sema::ActOnConditionalOp(SourceLocation QuestionLoc,
-    if (!(getLangOpts().CPlusPlus
-          && !commonExpr->isTypeDependent()
-          && commonExpr->getValueKind() == RHSExpr->getValueKind()
-          && commonExpr->isGLValue()
```
So I don't see how losing the above two value kind exceptions fix this?  Can you explain this better?

zwuis replied:
> So I don't see how losing the above two value kind exceptions fix this? Can you explain this better?

I have updated the description of this PR. Please let me know if it isn't still clear enough.

erichkeane commented on clang/test/SemaCXX/conditional-gnu-ext.cpp:
```diff
@@ -0,0 +1,19 @@
+// RUN: %clang_cc1 -fsyntax-only -verify -std=c++11 %s
+// RUN: %clang_cc1 -fsyntax-only -verify -std=c++11 %s -fexperimental-new-constant-interpreter
```
Should run with BOTH constexpr interpreters.

zwuis replied:
> Should run with BOTH constexpr interpreters.

Sorry I don't understand the meaning. Could you give an example?

erichkeane reviewed (commented):
I too am very concerned about the lack of tests here, there are some disappearing conditions I cannot help but think are there for a good reason, and would like to see some study as to why those where there and when.

I'm also concerned about the constexpr support being 'FIXME', it seems that this should just 'work' unless something odd is happening.

zwuis committed:
Fix previously undiscovered case and address review feedbacks

'conditional-gnu-exp.cpp' is mainly copied and modified from 'conditional-expr.cpp'
'conditional-gnu-ext.c' is mainly copied and modified from 'conditional-expr.c'

zwuis commented:
> I'm also concerned about the constexpr support being 'FIXME', it seems that this should just 'work' unless something odd is happening.

Sorry I forgot the restriction of constexpr function in C++11. Fixed.

zwuis committed:
Merge branch 'main' into fix#15998

zwuis committed:
Fix failed codegen test
